File: exchanges/huobi.py
from typing import Union

import requests


def get_huobi_spot_trade_link(currency1: str, currency2: str) -> str:
    return f"https://www.htx.com/ru-ru/trade/{currency1.lower()}_{currency2.lower()}?type=spot"


def get_huobi_futures_trade_link(currency1: str, currency2: str) -> str:
    return f"https://www.huobi.com/ru-ru/futures/linear_swap/exchange#contract_code={currency1}-{currency2}&contract_type=swap&type=isolated"


def get_huobi_funding_rates() -> dict[str, dict[str, Union[int, float]]]:
    funding_rates = requests.get(
        "https://api.hbdm.com/linear-swap-api/v1/swap_batch_funding_rate"
    )
    index_prices = requests.get("https://api.hbdm.com/linear-swap-api/v1/swap_index")
    mark_prices = requests.get(
        "https://api.hbdm.com/linear-swap-ex/market/detail/batch_merged"
    )

    return {
        funding_rate["contract_code"].replace("-", ""): {
            "funding_rate": float(funding_rate["funding_rate"]) * 100,
            "index_price": next(
                float(index_price["index_price"])
                for index_price in index_prices.json()["data"]
                if index_price["contract_code"] == funding_rate["contract_code"]
            ),
            "mark_price": next(
                float(mark_price["close"])
                for mark_price in mark_prices.json()["ticks"]
                if mark_price["contract_code"] == funding_rate["contract_code"]
            ),
            "next_funding_time": int(funding_rate["funding_time"]),
            "predicted_funding_rate": float(funding_rate["estimated_rate"]) * 100,
        }
        for funding_rate in funding_rates.json()["data"]
        if funding_rate["funding_rate"]
    }

File: exchanges/huobi.js
const axios = require('axios');

const { formatFundingRate, getFundingInterval, getTimeString } = require('../utils');

class Huobi {
  getSpotTradeLink(symbol) {
    const currency = symbol.split('USDT')[0].toLowerCase();
    return `https://www.htx.com/ru-ru/trade/${currency}_usdt?type=spot`;
  }

  getFuturesTradeLink(symbol) {
    const currency = symbol.split('USDT')[0];
    return `https://www.huobi.com/ru-ru/futures/linear_swap/exchange#contract_code=${currency}-USDT&contract_type=swap&type=cross`;
  }

  async getFundingRates() {
    try {
      const { data: fundingRates } = await axios.get('https://api.hbdm.com/linear-swap-api/v1/swap_batch_funding_rate');
      const { data: indexPrices } = await axios.get('https://api.hbdm.com/linear-swap-api/v1/swap_index');
      const { data: markPrices } = await axios.get('https://api.hbdm.com/linear-swap-ex/market/detail/batch_merged');

      return fundingRates.data
        .filter((fundingRate) => fundingRate.funding_rate)
        .reduce((acc, fundingRate) => {
          const symbol = fundingRate.contract_code.replace('-', '');

          const indexPrice = indexPrices.data.find(
            (indexPrice) => indexPrice.contract_code === fundingRate.contract_code
          ).index_price;

          const markPrice = markPrices.ticks.find(
            (markPrice) => markPrice.contract_code === fundingRate.contract_code
          ).close;

          const nextFundingTime = parseInt(fundingRate.funding_time);
          const fundingInterval = getFundingInterval(nextFundingTime, parseInt(fundingRate.next_funding_time));

          return {
            ...acc,
            [symbol.replace(/^10+/g, '')]: {
              fundingRate: formatFundingRate(fundingRate.funding_rate),
              indexPrice,
              markPrice,
              nextFundingTime: getTimeString(nextFundingTime),
              fundingInterval,
              predictedFundingRate: formatFundingRate(fundingRate.estimated_rate),
              spotLink: this.getSpotTradeLink(symbol),
              futuresLink: this.getFuturesTradeLink(symbol),
              multiplier: symbol.match(/^10+/g)?.[0] ?? 1,
            },
          };
        });
    } catch (err) {
      console.log(`Ошибка получения данных фандинга Huobi. ${err?.message}`);
    }
  }
}

module.exports = new Huobi();
